### PR TITLE
Work on supporting foreign calls

### LINKED ISSUE #277: concretization of `:foreigncall`
Testing with a reduced MWE from https://github.com/aviatesk/JET.jl/issues/224

```
using Libdl

let
    llvmpaths = filter(lib -> occursin(r"LLVM\b", basename(lib)), Libdl.dllist())
    if length(llvmpaths) != 1
        throw(ArgumentError("Found one or multiple LLVM libraries"))
    end
    libllvm = Libdl.dlopen(llvmpaths[1])
    gethostcpufeatures = Libdl.dlsym(libllvm, :LLVMGetHostCPUFeatures)
    features_cstring = ccall(gethostcpufeatures, Cstring, ())
    features = split(unsafe_string(features_cstring), ',')
    Libc.free(features_cstring)
    features
end
```

reveals

```
ERROR: BoundsError: attempt to access 31-element Vector{Bool} at index [32]
Stacktrace:
  [1] getindex
    @ .\array.jl:921 [inlined]
  [2] selective_eval!(recurse::Any, frame::JuliaInterpreter.Frame, isrequired::Vector{Bool}, istoplevel::Bool)
    @ LoweredCodeUtils C:\Users\Win10\.julia\packages\LoweredCodeUtils\jZY56\src\codeedges.jl:848
  [3] selective_eval_fromstart!
    @ C:\Users\Win10\.julia\packages\LoweredCodeUtils\jZY56\src\codeedges.jl:875 [inlined]
  [4] partially_interpret!(interp::JET.ConcreteInterpreter{JET.JETAnalyzer{JET.BasicPass{typeof(JET.basic_function_filter)}}, Expr}, mod::Module, src::Core.CodeInfo)
    @ JET C:\Users\Win10\Documents\GitHub\JET.jl\src\toplevel\virtualprocess.jl:799
  [5] _virtual_process!(toplevelex::Expr, filename::String, analyzer::JET.JETAnalyzer{JET.BasicPass{typeof(JET.basic_function_filter)}}, config::JET.ToplevelConfig{Expr}, context::Module, res::JET.VirtualProcessResult)
    @ JET C:\Users\Win10\Documents\GitHub\JET.jl\src\toplevel\virtualprocess.jl:630
  [6] _virtual_process!(s::String, filename::String, analyzer::JET.JETAnalyzer{JET.BasicPass{typeof(JET.basic_function_filter)}}, config::JET.ToplevelConfig{Expr}, context::Module, res::JET.VirtualProcessResult)
    @ JET C:\Users\Win10\Documents\GitHub\JET.jl\src\toplevel\virtualprocess.jl:432
  [7] virtual_process(x::String, filename::String, analyzer::JET.JETAnalyzer{JET.BasicPass{typeof(JET.basic_function_filter)}}, config::JET.ToplevelConfig{Expr})
    @ JET C:\Users\Win10\Documents\GitHub\JET.jl\src\toplevel\virtualprocess.jl:301
  [8] report_text(text::String, filename::String; analyzer::Type{JET.JETAnalyzer}, source::String, jetconfigs::Base.Pairs{Symbol, IOContext{Base.TTY}, Tuple{Symbol}, NamedTuple{(:toplevel_logger,), Tuple{IOContext{Base.TTY}}}})
    @ JET C:\Users\Win10\Documents\GitHub\JET.jl\src\JET.jl:1003
  [9] report_file(filename::String; __default_configs::Tuple{Pair{Symbol, IOContext{Base.TTY}}}, source::Nothing, jetconfigs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JET C:\Users\Win10\Documents\GitHub\JET.jl\src\JET.jl:826
 [10] report_file(filename::String)
    @ JET C:\Users\Win10\Documents\GitHub\JET.jl\src\JET.jl:801
 [11] top-level scope
    @ none:1
```

### PR DESCRIPTION
**Pull request by goerch**
Use optimized JuliaInterpreter frames to support foreign calls. Fixes #277 and #224, but breaks `eval` tests. Open question: should 

```
        analyzer = AbstractAnalyzer(analyzer, concretized, context)

        _, result = analyze_toplevel!(analyzer, src)
```

use `frame.framecode.src` instead to be synced with `concretized`? That change would break even more tests.

### CODE REVIEW

**Comment by goerch** [comment]
It seems our new problem has to do with 

```
            # analysis of `eval` calls are difficult, let's give up it and just evaluate
            # toplevel `eval` calls; they may contain toplevel definitions
            # adapted from
            # - https://github.com/timholy/Revise.jl/blob/266ed68d7dd3bea67c39f96513cda30bbcd7d441/src/lowered.jl#L53
            # - https://github.com/timholy/Revise.jl/blob/266ed68d7dd3bea67c39f96513cda30bbcd7d441/src/lowered.jl#L87-L88
            if begin
                    f === :eval ||
                    (callee_matches(f, Base, :getproperty) && is_quotenode_egal(stmt.args[end], :eval)) ||
                    isa(f, GlobalRef) && f.name === :eval
                end
                # statement `i` may be the equivalent of `f = Core.eval`, so require each
                # stmt that calls `eval` via `f(expr)`
                concretize[edges.succs[i]] .= true
                concretize[i] = true
            end
```

We probably do not detect one or more of these in the optimized code...

**Comment by aviatesk** [comment]
> Open question: should
> 
> ```
>         analyzer = AbstractAnalyzer(analyzer, concretized, context)
> 
>         _, result = analyze_toplevel!(analyzer, src)
> ```
> 
> use `frame.framecode.src` instead to be synced with `concretized`?

No, `analyze_toplevel` analyzes a lowered code that is already partially interpreted.

> We probably do not detect one or more of these in the optimized code...

Hm maype [this optimization](https://github.com/JuliaDebug/JuliaInterpreter.jl/blob/398721d10be2edf40883a11b98e210ccda5393e3/src/optimize.jl#L215-L240) breaks some invariant(s) that our `eval` handling assumes?

How about separating the `:foreigncall` handling from `JuliaInterpreter.optimize!` and probably enforce it within `Frame` construction (since it seems necessary for successful interpretation, rather than just to be a mere "optimization"), and disable other optimizations?

**Comment by goerch** [comment]
> > Open question: should
> > ```
> >         analyzer = AbstractAnalyzer(analyzer, concretized, context)
> > 
> >         _, result = analyze_toplevel!(analyzer, src)
> > ```
> >   
> > use `frame.framecode.src` instead to be synced with `concretized`?
> 
> No, `analyze_toplevel` analyzes a lowered code that is already partially interpreted.

OK, but I'm not sure if `length(concretized) == length(src)` holds before analyze_toplevel!. Will check with an assertion.

> Hm maype [this optimization](https://github.com/JuliaDebug/JuliaInterpreter.jl/blob/398721d10be2edf40883a11b98e210ccda5393e3/src/optimize.jl#L215-L240) breaks some invariant(s) that our `eval` handling assumes?

Probably.

> How about separating the `:foreigncall` handling from `JuliaInterpreter.optimize!` and probably enforce it within `Frame` construction (since it seems necessary for successful interpretation, rather than just to be a mere "optimization"), and disable other optimizations?

Maybe adapting `select_direct_requirement!` is good enough at the moment?

**Comment by goerch** [comment]
> OK, but I'm not sure if `length(concretized) == length(src.code)` holds before analyze_toplevel!. Will check with an assertion.

It doesn't.

**Comment by aviatesk** [comment]
> It doesn't hold. And `concretized` is accessed in the analyzer phase. This does not feel right.

Ah, I just know that `src::CodeInfo` is copied upon `FrameCode` construction. Yes, we need to make `src::CodeInfo` consistent with `concretized` even during abstract interpret.

Maybe we can do:
- tweak abstract interpret pass so that it can handle `:foreigncall` expressions that JuliaInterpreter emit
- keep using the original `src` for abstract interpret, with running `select_statements` twice (dirty)

**Comment by goerch** [comment]
> > It doesn't hold. And `concretized` is accessed in the analyzer phase. This does not feel right.
> 
> Ah, I just know that `src::CodeInfo` is copied upon `FrameCode` construction. Yes, we need to make `src::CodeInfo` consistent with `concretized` even during abstract interpret.
> 
> Maybe we can do:
> 
>     * tweak abstract interpret pass so that it can handle `:foreigncall` expressions that JuliaInterpreter emit
> 
>     * keep using the original `src` for abstract interpret, with running `select_statements` twice (dirty)

:) This does the trick!

**Comment by goerch** [comment]
Now I only need some outline where to add which tests?

**Comment by aviatesk** [comment]
Anywhere you want. Maybe test_virtualprocess.jl?

**Comment by goerch** [comment]
Just noticed: we better keep two versions of select_statements around?

**Comment by aviatesk** [comment]
How about switching the behaviors by new `optimized::Bool` flag?

**Comment by goerch** [comment]
OK, `ScanByte` is go I hope, now I'll have to look into `Convex` still failing with 

```
┌ @ C:\Users\Win10\.julia\packages\Convex\uI27T\src\variable.jl:204 
│ UndefRefError: access to undefined reference
└───────────────────────────────────────────────────────────────────
```

**Comment by goerch** [comment]
Test failure due to 

```
   Evaluated: isempty(ToplevelErrorReport[ActualErrorWrapped(ArgumentError("Package Libdl not found in current path:\n- Run `import Pkg; Pkg.add(\"Libdl\")` to install the Libdl package.\n"), 
```

Hm. Do we need to set this up in CI somehow?

File: src/toplevel/virtualprocess.jl [hunk 1]
```diff
@@ -627,11 +627,17 @@ function _virtual_process!(toplevelex::Expr,
...
+        # generate optimized code to support foreign calls,
+        # see https://github.com/JuliaDebug/JuliaInterpreter.jl/issues/13
+        frame = Frame(context, src)
+        concretized = partially_interpret!(interp, frame)
 
         # bail out if nothing to analyze (just a performance optimization)
         all(concretized) && continue
 
+        # concretize the unoptimized source for the analyzer
+        concretize = select_statements(src, false)
```
Comment (aviatesk) [inline src/toplevel/virtualprocess.jl:639]:
```suggestion
        concretized = select_statements(src, false)
```

**Comment by aviatesk** [comment]
> Do we need to set this up in CI somehow?

https://pkgdocs.julialang.org/v1/creating-packages/#Test-specific-dependencies-in-Julia-1.2-and-above

File: src/toplevel/virtualprocess.jl [hunk 2]
```diff
@@ -627,11 +627,17 @@ function _virtual_process!(toplevelex::Expr,
                                      config,
                                      res,
                                      )
-        concretized = partially_interpret!(interp, context, src)
+        # generate optimized code to support foreign calls,
+        # see https://github.com/JuliaDebug/JuliaInterpreter.jl/issues/13
+        frame = Frame(context, src)
```
Comment (aviatesk) [inline src/toplevel/virtualprocess.jl:632]: Now this part can be put back into `partially_interpret!`?

File: src/toplevel/virtualprocess.jl [hunk 3]
```diff
@@ -627,11 +627,17 @@ function _virtual_process!(toplevelex::Expr,
...
+        concretized = partially_interpret!(interp, frame)
 
         # bail out if nothing to analyze (just a performance optimization)
         all(concretized) && continue
 
+        # concretize the unoptimized source for the analyzer
```
Comment (aviatesk) [inline src/toplevel/virtualprocess.jl:638]:
```suggestion
        # recalculate concretizations for the original unoptimized source for abstract interpretation
        # NOTE JuliaInterpreter might have produced "optimized" `:foreigncall` expressions that 
        # can't be handled by Julia's native abstract interpretation routine
```

File: test/toplevel/test_virtualprocess.jl [hunk 4]
```diff
@@ -1022,6 +1022,21 @@ end
...
+        res = @analyze_toplevel begin
+            using Libdl
+            llvmpaths = filter(lib -> occursin(r"LLVM\b", basename(lib)), Libdl.dllist())
+            if length(llvmpaths) != 1
+                throw(ArgumentError("Found one or multiple LLVM libraries"))
+            end
+            libllvm = Libdl.dlopen(llvmpaths[1])
+            gethostcpufeatures = Libdl.dlsym(libllvm, :LLVMGetHostCPUFeatures)
+            ccall(gethostcpufeatures, Cstring, ())
+        end
```
Comment (aviatesk) [inline test/toplevel/test_virtualprocess.jl:1036]: This top-level code is actually abstract-interpreted:
```julia
julia> res = @analyze_toplevel toplevel_logger=IOContext(stdout, :JET_LOGGER_LEVEL=>1) begin
           using Libdl
           llvmpaths = filter(lib -> occursin(r"LLVM\b", basename(lib)), Libdl.dllist())
           if length(llvmpaths) != 1
               throw(ArgumentError("Found one or multiple LLVM libraries"))
           end
           libllvm = Libdl.dlopen(llvmpaths[1])
           gethostcpufeatures = Libdl.dlsym(libllvm, :LLVMGetHostCPUFeatures)
           ccall(gethostcpufeatures, Cstring, ())
       end
[toplevel-debug] virtualized the context of Main (took 0.066 sec)
[toplevel-debug] concretization plan at REPL[2]:2:
1 t 1 ─     using Libdl
2 f └──     return nothing
[toplevel-debug] concretization plan at REPL[2]:3:
 1 t 1 ─       $(Expr(:thunk, CodeInfo(
    @ none within `top-level scope`
1 ─      global var"#1#2"
│        const var"#1#2"
│   %3 = Core._structtype(Main.##JETVirtualModule#278, Symbol("#1#2"), Core.svec(), Core.svec(), Core.svec(), false, 0)
│        var"#1#2" = %3
│        Core._setsuper!(var"#1#2", Core.Function)
│        Core._typebody!(var"#1#2", Core.svec())
└──      return nothing
)))
 2 t │   %2  = Core.svec(var"#1#2", Core.Any)
 3 t │   %3  = Core.svec()
 4 t │   %4  = Core.svec(%2, %3, $(QuoteNode(:(#= REPL[2]:3 =#))))
 5 t │         $(Expr(:method, false, :(%4), CodeInfo(
    @ REPL[2]:3 within `none`
1 ─ %1 = basename(lib)
│   %2 = occursin(r"LLVM\b", %1)
└──      return %2
)))
 6 f │         _1 = %new(var"#1#2")
 7 f │   %7  = _1
 8 f │   %8  = Base.getproperty(Libdl, :dllist)
 9 f │   %9  = (%8)()
10 f │   %10 = filter(%7, %9)
11 f │         llvmpaths = %10
12 f └──       return %10
[toplevel-debug] concretization plan at REPL[2]:4:
1 f 1 ─ %1 = length(llvmpaths)
2 f │   %2 = %1 != 1
3 f └──      goto #3 if not %2
4 f 2 ─ %4 = ArgumentError("Found one or multiple LLVM libraries")
5 f │   %5 = throw(%4)
6 f └──      return %5
7 f 3 ─      return nothing
[toplevel-debug] concretization plan at REPL[2]:7:
1 f 1 ─ %1 = Base.getproperty(Libdl, :dlopen)
2 f │   %2 = Base.getindex(llvmpaths, 1)
3 f │   %3 = (%1)(%2)
4 f │        libllvm = %3
5 f └──      return %3
[toplevel-debug] concretization plan at REPL[2]:8:
1 f 1 ─ %1 = Base.getproperty(Libdl, :dlsym)
2 f │   %2 = (%1)(libllvm, :LLVMGetHostCPUFeatures)
3 f │        gethostcpufeatures = %2
4 f └──      return %2
[toplevel-debug] concretization plan at REPL[2]:9:
1 f 1 ─ %1 = $(Expr(:foreigncall, :gethostcpufeatures, :Cstring, :(Core.svec()), 0, :(:ccall)))
2 f └──      return %1
```
(and thus this testcase passes without this PR)

Comment (aviatesk) [reply test/toplevel/test_virtualprocess.jl:1036]: After a closer look, it seems like the root problem here is not the missing support for `:foreigncall` expressions, but inappropriate statement selections.
I could successfully analyze `ScanByte` by this configuration: `report_package("ScanByte", toplevel_logger=IOContext(stdout, :JET_LOGGER_LEVEL=>1), concretization_patterns=[:(const x_ = y_), :(let; x_; end)])`

File: src/toplevel/virtualprocess.jl (hunk 2 above)
Comment (goerch) [reply src/toplevel/virtualprocess.jl:632]: Done.

File: src/toplevel/virtualprocess.jl (hunk 3 above)
Comment (goerch) [reply src/toplevel/virtualprocess.jl:638]: Well explained!

File: test/toplevel/test_virtualprocess.jl (hunk 4 above)
Comment (goerch) [reply test/toplevel/test_virtualprocess.jl:1038]: Sorry, needed to wrap the code into a (let) block.

Comment (aviatesk) [reply test/toplevel/test_virtualprocess.jl:1038]: Yeah, well, my point was that the latest master will pass the test case even when it's wrapped into `let` block -- I think it implies `:foreigncall` expression tweak is _not_ necessary for concretization by JuliaInterpreter.

**Comment by aviatesk** [comment]
Alright, now #224 and #281 are fixed by #283, but still I'd like to merge some parts of this PR.
How about:
1. first rebasing this PR against #283 
2. removing changes around letting JuliaInterpreter optimize `src::CodeInfo`
3. adding new assertion that makes sure `concretize` is consistent with `src` that abstract interpreter works on
4. keeping `test/Project.toml` changes
?

**Comment by goerch** [comment]
> Alright, now #224 and #281 are fixed by #283, but still I'd like to merge some parts of this PR. How about:
> 
>     1. first rebasing this PR against [concretize all top-level code when analyzing a package #283](https://github.com/aviatesk/JET.jl/pull/283)

Done.

>     2. removing changes around letting JuliaInterpreter optimize `src::CodeInfo`
> 
>     3. adding new assertion that makes sure `concretize` is consistent with `src` that abstract interpreter works on

This assertion still leads still to test errors?

>     4. keeping `test/Project.toml` changes

OK.

**Comment by aviatesk** [comment]
Could you explain why we need to use `optimize` option of JuliaInterpreter?

**Comment by goerch** [comment]
> Could you explain why we need to use `optimize` option of JuliaInterpreter?

We don't need to, it is the default (which is not changeable in the default `Frame` constructor AFAIU). One of my older patches constructed the `Frame` therefore by hand using a copy of `JuliaInterpreter` internal code: https://github.com/aviatesk/JET.jl/pull/280/commits/618b814714bf8697379dd19dc445cecfc8e46af1#diff-a8508c23a55035016aa42d51c599c9ce18ff31cc603ad173d0612849e35e457dL799-L803

It is your decision, which implementation is more future proof.

**Comment by aviatesk** [comment]
I'd like to avoid duplicated calculation of `select_statements` since there is no strong guarantee of correspondance between statements concritized by `ConcreteInterpreter` and selected statements obtained by the recalculation.

**Comment by goerch** [comment]
> I'd like to avoid duplicated calculation of `select_statements` since there is no strong guarantee of correspondance between statements concritized by `ConcreteInterpreter` and selected statements obtained by the recalculation.

Will change and check.

**Comment by aviatesk** [comment]
We can also add `optimize` option to `JuliaInterpreter.Frame` constructor :)

**Review by aviatesk — APPROVED** [review]
Thanks for working on this.

**Pull request merged**